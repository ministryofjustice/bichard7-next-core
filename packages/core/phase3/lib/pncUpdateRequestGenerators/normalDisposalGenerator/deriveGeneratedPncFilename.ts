import type { HearingDefendant } from "../../../../types/AnnotatedHearingOutcome"

import { GENERATED_PNC_FILENAME_MAX_LENGTH } from "../../../../phase1/enrichAho/enrichFunctions/enrichDefendant/enrichDefendant"

const ILLEGAL_FILENAME_PATTERN = new RegExp("[^a-zA-Z0-9\\- /]", "g")

const deriveGeneratedPncFilename = (defendant: HearingDefendant) => {
  if (defendant.DefendantDetail != null) {
    const generatedPNCFilename = defendant.DefendantDetail?.GeneratedPNCFilename ?? ""
    return generatedPNCFilename.includes("/") ? generatedPNCFilename : `${generatedPNCFilename}/`
  }

  let generatedPNCFilename = defendant.OrganisationName?.replace(ILLEGAL_FILENAME_PATTERN, " ")
    .replace(/\s+/g, " ")
    .trim()

  if (!generatedPNCFilename) {
    return ""
  }

  if (generatedPNCFilename.length > GENERATED_PNC_FILENAME_MAX_LENGTH) {
    generatedPNCFilename = generatedPNCFilename.substring(0, GENERATED_PNC_FILENAME_MAX_LENGTH - 1) + "+"
  }

  if (!generatedPNCFilename.includes("/")) {
    if (generatedPNCFilename.length == GENERATED_PNC_FILENAME_MAX_LENGTH) {
      generatedPNCFilename = generatedPNCFilename.substring(0, GENERATED_PNC_FILENAME_MAX_LENGTH - 2) + "/+"
    } else {
      generatedPNCFilename = generatedPNCFilename + "/"
    }
  }

  return generatedPNCFilename
}

export default deriveGeneratedPncFilename
