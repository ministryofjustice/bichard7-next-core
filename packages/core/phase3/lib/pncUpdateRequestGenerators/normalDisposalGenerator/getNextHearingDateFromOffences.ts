import type { Offence } from "@moj-bichard7/common/types/AnnotatedHearingOutcome"

const getNextHearingDateFromOffences = (offences: Offence[]): Date | undefined => {
  const nextHearingDate =
    offences.flatMap((offence) => offence.Result).find((result) => result.PNCDisposalType === 2059)?.NextHearingDate ??
    undefined

  return nextHearingDate ? new Date(nextHearingDate) : undefined
}

export default getNextHearingDateFromOffences
