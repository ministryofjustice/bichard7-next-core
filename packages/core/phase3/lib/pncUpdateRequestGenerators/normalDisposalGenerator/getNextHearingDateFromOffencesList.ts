import type { Offence } from "../../../../types/AnnotatedHearingOutcome"

const getNextHearingDateFromOffencesList = (offences: Offence[]): Date | undefined => {
  const nextHearingDate =
    offences.flatMap((offence) => offence.Result).find((result) => result.PNCDisposalType === 2059)?.NextHearingDate ??
    undefined

  return nextHearingDate ? new Date(nextHearingDate) : undefined
}

export default getNextHearingDateFromOffencesList
