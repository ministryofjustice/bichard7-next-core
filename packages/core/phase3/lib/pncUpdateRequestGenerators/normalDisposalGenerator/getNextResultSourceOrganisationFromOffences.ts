import type { Offence, OrganisationUnitCodes } from "@moj-bichard7/common/types/AnnotatedHearingOutcome"

const getNextResultSourceOrganisationFromOffences = (offences: Offence[]): OrganisationUnitCodes | undefined =>
  offences
    .flatMap((offence) => offence.Result)
    .find((result) => result.NextResultSourceOrganisation && result.PNCDisposalType === 2059)
    ?.NextResultSourceOrganisation ?? undefined

export default getNextResultSourceOrganisationFromOffences
