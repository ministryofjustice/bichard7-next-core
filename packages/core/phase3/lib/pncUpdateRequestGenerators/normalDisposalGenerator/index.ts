export { default as normalDisposalGenerator } from "./normalDisposalGenerator"
