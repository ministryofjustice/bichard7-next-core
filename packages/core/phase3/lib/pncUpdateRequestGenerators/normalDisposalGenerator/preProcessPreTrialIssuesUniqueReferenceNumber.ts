const preProcessPreTrialIssuesUniqueReferenceNumber = (ptiUrn?: string, forceOwner?: string) => {
  const forceCode = (forceOwner?.length === 6 ? forceOwner : ptiUrn)?.substring(0, 4) ?? ""
  return [forceCode.padStart(4, " "), "/", ptiUrn?.substring(4, 18)].join("")
}

export default preProcessPreTrialIssuesUniqueReferenceNumber
