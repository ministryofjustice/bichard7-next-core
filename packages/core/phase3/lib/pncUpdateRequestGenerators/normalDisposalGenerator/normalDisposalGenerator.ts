import type { Result } from "@moj-bichard7/common/types/Result"

import { isError } from "@moj-bichard7/common/types/Result"

import type { PncUpdateArrestHearingAdjudicationAndDisposal } from "../../../types/HearingDetails"
import type PncUpdateRequestGenerator from "../../../types/PncUpdateRequestGenerator"

import formatDateSpecifiedInResult from "../../../../lib/createPncDisposalsFromResult/formatDateSpecifiedInResult"
import getAdjustedRecordableOffencesForCourtCase from "../../../../lib/getAdjustedRecordableOffencesForCourtCase"
import isResultCompatibleWithDisposal from "../../../../phase2/lib/isResultCompatibleWithDisposal"
import { PncOperation } from "../../../../types/PncOperation"
import generateBasePncUpdateRequest from "../../generateBasePncUpdateRequest"
import getPncCourtCode from "../../getPncCourtCode"
import { generateArrestHearingsAdjudicationsAndDisposals } from "../../hearingDetails/generateArrestHearingsAdjudicationsAndDisposals"
import { generateHearingsAdjudicationsAndDisposals } from "../../hearingDetails/generateHearingsAdjudicationsAndDisposals"
import preProcessAsn from "../../preProcessAsn"
import preProcessCourtCaseReferenceNumber from "../../preProcessCourtCaseReferenceNumber"
import deriveGeneratedPncFilename from "./deriveGeneratedPncFilename"
import getNextHearingDateFromOffences from "./getNextHearingDateFromOffences"
import getNextResultSourceOrganisationFromOffences from "./getNextResultSourceOrganisationFromOffences"
import preProcessPreTrialIssuesUniqueReferenceNumber from "./preProcessPreTrialIssuesUniqueReferenceNumber"

const COURT_CODE_WHEN_DEFENDANT_FAILED_TO_APPEAR = "9998"
const COURT_TYPE_NOT_AVAILABLE = "Not available from Court"

const preProcessCourtCode = (courtCode: string) => courtCode.padStart(4, "0")

const normalDisposalGenerator: PncUpdateRequestGenerator<PncOperation.NORMAL_DISPOSAL> = (
  pncUpdateDataset,
  operation
) => {
  const hearing = pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Hearing
  const hearingCase = pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case
  const hearingDefendant = hearingCase.HearingDefendant
  const offences = getAdjustedRecordableOffencesForCourtCase(
    hearingDefendant.Offence,
    operation.data?.courtCaseReference
  )

  const courtCaseReference = operation.data?.courtCaseReference ?? hearingCase.CourtCaseReferenceNumber
  const formattedCourtCaseReference = preProcessCourtCaseReferenceNumber(courtCaseReference)
  if (isError(formattedCourtCaseReference)) {
    return formattedCourtCaseReference
  }

  const couPsaCourtCode = getPncCourtCode(hearing.CourtHearingLocation, hearing.CourtHouseCode)
  if (isError(couPsaCourtCode)) {
    return couPsaCourtCode
  }

  const nextResultSourceOrganisation = getNextResultSourceOrganisationFromOffences(offences)
  const crtPsaCourtCode = getPncCourtCode(nextResultSourceOrganisation, hearing.CourtHouseCode)
  if (isError(crtPsaCourtCode)) {
    return crtPsaCourtCode
  }

  const courtDate = crtPsaCourtCode ? getNextHearingDateFromOffences(offences) : undefined
  const pendingCourtHouseName =
    crtPsaCourtCode === COURT_CODE_WHEN_DEFENDANT_FAILED_TO_APPEAR ? COURT_TYPE_NOT_AVAILABLE : ""
  const hearingsAdjudicationsAndDisposals = generateHearingsAdjudicationsAndDisposals(
    pncUpdateDataset,
    operation.data?.courtCaseReference
  )
  let arrestSummonsNumber: Result<null | string> = null
  let arrestsAdjudicationsAndDisposals: PncUpdateArrestHearingAdjudicationAndDisposal[] = []
  const hasOffencesAddedByTheCourt = offences.some(
    (offence) => offence.AddedByTheCourt && isResultCompatibleWithDisposal(offence)
  )
  if (hasOffencesAddedByTheCourt) {
    arrestSummonsNumber = preProcessAsn(hearingDefendant.ArrestSummonsNumber)
    if (isError(arrestSummonsNumber)) {
      return arrestSummonsNumber
    }

    arrestsAdjudicationsAndDisposals = generateArrestHearingsAdjudicationsAndDisposals(
      pncUpdateDataset,
      operation.data?.courtCaseReference
    )
  }

  return {
    operation: PncOperation.NORMAL_DISPOSAL,
    request: {
      ...generateBasePncUpdateRequest(pncUpdateDataset),
      arrestSummonsNumber,
      arrestsAdjudicationsAndDisposals,
      courtCaseReferenceNumber: formattedCourtCaseReference,
      courtHouseName:
        couPsaCourtCode === COURT_CODE_WHEN_DEFENDANT_FAILED_TO_APPEAR
          ? `${hearing.CourtHouseName} ${hearing.CourtType}`
          : "",
      dateOfHearing: formatDateSpecifiedInResult(hearing.DateOfHearing, true),
      generatedPNCFilename: deriveGeneratedPncFilename(hearingDefendant),
      hearingsAdjudicationsAndDisposals,
      pendingCourtDate: courtDate ? formatDateSpecifiedInResult(courtDate, true) : null,
      pendingCourtHouseName: crtPsaCourtCode ? pendingCourtHouseName : null,
      pendingPsaCourtCode: crtPsaCourtCode ? preProcessCourtCode(crtPsaCourtCode) : null,
      preTrialIssuesUniqueReferenceNumber: preProcessPreTrialIssuesUniqueReferenceNumber(
        offences,
        courtCaseReference,
        hearingCase.PTIURN,
        hearingCase.ForceOwner?.OrganisationUnitCode ?? undefined
      ),
      psaCourtCode: couPsaCourtCode ? preProcessCourtCode(couPsaCourtCode) : null
    }
  }
}

export default normalDisposalGenerator
