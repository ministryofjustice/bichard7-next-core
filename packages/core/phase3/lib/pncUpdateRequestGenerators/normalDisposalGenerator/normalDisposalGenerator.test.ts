import { isError } from "@moj-bichard7/common/types/Result"

import type { Offence, OffenceReason, Result } from "../../../../types/AnnotatedHearingOutcome"
import type { Operation, PncUpdateDataset } from "../../../../types/PncUpdateDataset"
import type NormalDisposalPncUpdateRequest from "../../../types/NormalDisposalPncUpdateRequest"

import lookupOrganisationUnitByCode from "../../../../lib/dataLookup/lookupOrganisationUnitByCode"
import { PncOperation } from "../../../../types/PncOperation"
import ResultClass from "../../../../types/ResultClass"
import normalDisposalGenerator from "./normalDisposalGenerator"

jest.mock("../../../../lib/dataLookup/lookupOrganisationUnitByCode")
const mockedLookupOrganisationUnitByCode = lookupOrganisationUnitByCode as jest.Mock

const createPncUpdateDataset = () => {
  const offence = {
    OffenceCategory: "ZZ",
    CriminalProsecutionReference: {
      OffenceReason: {
        __type: "NationalOffenceReason",
        OffenceCode: {
          __type: "NonMatchingOffenceCode",
          ActOrSource: "RT",
          Year: "88",
          Reason: 191,
          FullCode: "RT88191"
        }
      } as unknown as OffenceReason
    },
    AddedByTheCourt: true,
    ActualOffenceStartDate: {
      StartDate: new Date("2025-01-01")
    },
    Result: [
      {
        NextHearingDate: "2024-12-11T10:11:12.000Z",
        NextResultSourceOrganisation: {
          TopLevelCode: "B",
          SecondLevelCode: "01",
          ThirdLevelCode: "00",
          BottomLevelCode: "00",
          OrganisationUnitCode: "B000000"
        },
        PNCDisposalType: 2059,
        PNCAdjudicationExists: true,
        ResultClass: ResultClass.ADJOURNMENT_PRE_JUDGEMENT,
        ResultQualifierVariable: [{ Code: "LE" }]
      }
    ] as Result[]
  } as Offence

  const pncUpdateDataset = {
    AnnotatedHearingOutcome: {
      HearingOutcome: {
        Hearing: {
          DateOfHearing: new Date("2024-12-05"),
          CourtHearingLocation: {
            TopLevelCode: "B",
            SecondLevelCode: "01",
            ThirdLevelCode: "00",
            BottomLevelCode: "00",
            OrganisationUnitCode: "B000000"
          },
          CourtHouseName: "Magistrates' Courts London Croydon",
          CourtType: "MCA"
        },
        Case: {
          CourtCaseReferenceNumber: "97/1626/008395Q",
          ForceOwner: {
            TopLevelCode: "A",
            SecondLevelCode: "02",
            ThirdLevelCode: "BJ",
            BottomLevelCode: "01",
            OrganisationUnitCode: "A02BJ01"
          },
          HearingDefendant: {
            ArrestSummonsNumber: "1101ZD0100000410780J",
            BailConditions: ["This is a dummy bail condition."],
            Offence: [offence],
            RemandStatus: "CB"
          },
          PTIURN: "01ZD0303208"
        }
      }
    },
    Exceptions: [],
    PncOperations: []
  } as unknown as PncUpdateDataset

  return pncUpdateDataset
}

const operation = {
  code: "DISARR",
  data: {
    courtCaseReference: "97/1626/008395Q"
  }
} as Operation<PncOperation.NORMAL_DISPOSAL>

describe("normalDisposalGenerator", () => {
  beforeEach(() => {
    mockedLookupOrganisationUnitByCode.mockRestore()
  })

  it("should return error when ASN is invalid", () => {
    const pncUpdateDataset = createPncUpdateDataset()
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.ArrestSummonsNumber =
      "1101ZD0100000410780JXX"

    const result = normalDisposalGenerator(pncUpdateDataset, operation)

    expect(isError(result)).toBe(true)
    expect((result as Error).message).toBe("Invalid ASN length. Length is 22")
  })

  it("should return error when court case reference in operation is invalid", () => {
    const pncUpdateDataset = createPncUpdateDataset()
    const operationWithInvalidCourtCaseReference = {
      code: PncOperation.NORMAL_DISPOSAL,
      data: {
        courtCaseReference: "97/1626/008395"
      }
    } as Operation<PncOperation.NORMAL_DISPOSAL>

    const result = normalDisposalGenerator(pncUpdateDataset, operationWithInvalidCourtCaseReference)

    expect(isError(result)).toBe(true)
    expect((result as Error).message).toBe("Court Case Reference Number length must be 15, but the length is 14")
  })

  it("should return error when court case reference in operation does not exist and court case reference in hearing outcome is invalid", () => {
    const pncUpdateDataset = createPncUpdateDataset()
    const operationWithoutCourtCaseReference = {
      code: PncOperation.NORMAL_DISPOSAL,
      data: undefined
    } as Operation<PncOperation.NORMAL_DISPOSAL>
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.CourtCaseReferenceNumber = "97/1626/00839"

    const result = normalDisposalGenerator(pncUpdateDataset, operationWithoutCourtCaseReference)

    expect(isError(result)).toBe(true)
    expect((result as Error).message).toBe("Court Case Reference Number length must be 15, but the length is 13")
  })

  it("should return error when it fails to get PSA court code from the court hearing location", () => {
    const mockedLookupOrganisationUnitByCode = lookupOrganisationUnitByCode as jest.Mock

    mockedLookupOrganisationUnitByCode.mockReturnValue({
      bottomLevelCode: "00",
      bottomLevelName: "",
      secondLevelCode: "20",
      secondLevelName: "West Midlands",
      thirdLevelCode: "BN",
      thirdLevelName: "Birmingham Youth Court (Steelehouse Lane)",
      thirdLevelPsaCode: "I'm not a number",
      topLevelCode: "B",
      topLevelName: "Magistrates' Courts"
    })

    const pncUpdateDataset = createPncUpdateDataset()
    const operationWithoutCourtCaseReference = {
      code: PncOperation.NORMAL_DISPOSAL,
      data: undefined
    } as Operation<PncOperation.NORMAL_DISPOSAL>
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Hearing.CourtHouseCode = 4001

    const result = normalDisposalGenerator(pncUpdateDataset, operationWithoutCourtCaseReference)

    expect(isError(result)).toBe(true)
    expect((result as Error).message).toBe("PSA code 'I'm not a number' is not a number")
  })

  it("should return error when it fails to get PSA court code from the next result source organisation", () => {
    mockedLookupOrganisationUnitByCode.mockReturnValueOnce({
      bottomLevelCode: "00",
      bottomLevelName: "",
      secondLevelCode: "20",
      secondLevelName: "West Midlands",
      thirdLevelCode: "BN",
      thirdLevelName: "Birmingham Youth Court (Steelehouse Lane)",
      thirdLevelPsaCode: "2525",
      topLevelCode: "B",
      topLevelName: "Magistrates' Courts"
    })

    mockedLookupOrganisationUnitByCode.mockReturnValueOnce({
      bottomLevelCode: "00",
      bottomLevelName: "",
      secondLevelCode: "20",
      secondLevelName: "West Midlands",
      thirdLevelCode: "BN",
      thirdLevelName: "Birmingham Youth Court (Steelehouse Lane)",
      thirdLevelPsaCode: "I'm not a number 2",
      topLevelCode: "B",
      topLevelName: "Magistrates' Courts"
    })

    const pncUpdateDataset = createPncUpdateDataset()
    const operationWithoutCourtCaseReference = {
      code: PncOperation.NORMAL_DISPOSAL,
      data: undefined
    } as Operation<PncOperation.NORMAL_DISPOSAL>
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Hearing.CourtHouseCode = 4001

    const result = normalDisposalGenerator(pncUpdateDataset, operationWithoutCourtCaseReference)

    expect(isError(result)).toBe(true)
    expect((result as Error).message).toBe("PSA code 'I'm not a number 2' is not a number")
  })

  it("should not set ASN and arrests adjudications and disposals when there are no offences added by the court with results compatible with disposals", () => {
    const pncUpdateDataset = createPncUpdateDataset()
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.Offence[0].AddedByTheCourt = false

    const result = normalDisposalGenerator(pncUpdateDataset, operation)

    expect((result as NormalDisposalPncUpdateRequest).request.arrestSummonsNumber).toBeNull()
    expect(result).toMatchSnapshot()
  })

  it("should set ASN and arrests adjudications and disposals when there is an offence added by the court with results compatible with disposals", () => {
    const pncUpdateDataset = createPncUpdateDataset()

    const result = normalDisposalGenerator(pncUpdateDataset, operation)

    expect((result as NormalDisposalPncUpdateRequest).request.arrestSummonsNumber).toBe("11/01ZD/01/410780J")
    expect(result).toMatchSnapshot()
  })

  it("should set PTIURN (Pre Trial Issues Unique Reference Number) when case requires RCC and has reportable offences", () => {
    const pncUpdateDataset = createPncUpdateDataset()
    const offence = pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.Offence[0]
    const offence2 = structuredClone(offence)
    offence2.Result[0].PNCDisposalType = 2060
    const offence3 = structuredClone(offence)
    offence3.Result[0].PNCDisposalType = 2061
    pncUpdateDataset.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.Offence.push(offence2, offence3)

    const result = normalDisposalGenerator(pncUpdateDataset, operation)

    expect((result as NormalDisposalPncUpdateRequest).request.preTrialIssuesUniqueReferenceNumber).toBe("01ZD/0303208")
    expect(result).toMatchSnapshot()
  })
})
