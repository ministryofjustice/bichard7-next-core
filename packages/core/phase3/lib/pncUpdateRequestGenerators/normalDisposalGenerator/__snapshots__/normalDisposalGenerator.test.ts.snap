// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalDisposalGenerator should not set ASN and arrests adjudications and disposals when there are no offences added by the court with results compatible with disposals 1`] = `
{
  "operation": "DISARR",
  "request": {
    "arrestSummonsNumber": null,
    "arrestsAdjudicationsAndDisposals": [],
    "courtCaseReferenceNumber": "97/1626/8395Q",
    "courtHouseName": "",
    "croNumber": null,
    "dateOfHearing": "05122024",
    "forceStationCode": "02YZ",
    "generatedPNCFilename": "",
    "hearingsAdjudicationsAndDisposals": [
      {
        "courtOffenceSequenceNumber": "",
        "offenceReason": "RT88191",
        "type": "ORDINARY",
      },
      {
        "hearingDate": "",
        "numberOffencesTakenIntoAccount": "0000",
        "pleaStatus": "",
        "type": "ADJUDICATION",
        "verdict": "",
      },
      {
        "disposalQualifiers": "",
        "disposalQuantity": "                      00",
        "disposalText": "",
        "disposalType": "2059",
        "type": "DISPOSAL",
      },
    ],
    "pendingCourtDate": null,
    "pendingCourtHouseName": null,
    "pendingPsaCourtCode": null,
    "pncCheckName": null,
    "pncIdentifier": null,
    "preTrialIssuesUniqueReferenceNumber": null,
    "psaCourtCode": null,
  },
}
`;

exports[`normalDisposalGenerator should set ASN and arrests adjudications and disposals when there is an offence added by the court with results compatible with disposals 1`] = `
{
  "operation": "DISARR",
  "request": {
    "arrestSummonsNumber": "11/01ZD/01/410780J",
    "arrestsAdjudicationsAndDisposals": [
      {
        "committedOnBail": "N",
        "courtOffenceSequenceNumber": null,
        "locationOfOffence": "Not provided by Court",
        "offenceEndDate": "",
        "offenceEndTime": "",
        "offenceLocationFSCode": "02BJ",
        "offenceReason": "RT88191",
        "offenceReasonSequence": "",
        "offenceStartDate": "01012025",
        "offenceStartTime": "",
        "type": "ARREST",
      },
      {
        "hearingDate": "",
        "numberOffencesTakenIntoAccount": "0000",
        "pleaStatus": "",
        "type": "ADJUDICATION",
        "verdict": "",
      },
      {
        "disposalQualifiers": "",
        "disposalQuantity": "                      00",
        "disposalText": "",
        "disposalType": "2059",
        "type": "DISPOSAL",
      },
    ],
    "courtCaseReferenceNumber": "97/1626/8395Q",
    "courtHouseName": "",
    "croNumber": null,
    "dateOfHearing": "05122024",
    "forceStationCode": "02YZ",
    "generatedPNCFilename": "",
    "hearingsAdjudicationsAndDisposals": [],
    "pendingCourtDate": null,
    "pendingCourtHouseName": null,
    "pendingPsaCourtCode": null,
    "pncCheckName": null,
    "pncIdentifier": null,
    "preTrialIssuesUniqueReferenceNumber": null,
    "psaCourtCode": null,
  },
}
`;

exports[`normalDisposalGenerator should set PTIURN (Pre Trial Issues Unique Reference Number) when case requires RCC and has reportable offences 1`] = `
{
  "operation": "DISARR",
  "request": {
    "arrestSummonsNumber": "11/01ZD/01/410780J",
    "arrestsAdjudicationsAndDisposals": [
      {
        "committedOnBail": "N",
        "courtOffenceSequenceNumber": null,
        "locationOfOffence": "Not provided by Court",
        "offenceEndDate": "",
        "offenceEndTime": "",
        "offenceLocationFSCode": "02BJ",
        "offenceReason": "RT88191",
        "offenceReasonSequence": "",
        "offenceStartDate": "01012025",
        "offenceStartTime": "",
        "type": "ARREST",
      },
      {
        "hearingDate": "",
        "numberOffencesTakenIntoAccount": "0000",
        "pleaStatus": "",
        "type": "ADJUDICATION",
        "verdict": "",
      },
      {
        "disposalQualifiers": "",
        "disposalQuantity": "                      00",
        "disposalText": "",
        "disposalType": "2059",
        "type": "DISPOSAL",
      },
      {
        "committedOnBail": "N",
        "courtOffenceSequenceNumber": null,
        "locationOfOffence": "Not provided by Court",
        "offenceEndDate": "",
        "offenceEndTime": "",
        "offenceLocationFSCode": "02BJ",
        "offenceReason": "RT88191",
        "offenceReasonSequence": "",
        "offenceStartDate": "01012025",
        "offenceStartTime": "",
        "type": "ARREST",
      },
      {
        "hearingDate": "",
        "numberOffencesTakenIntoAccount": "0000",
        "pleaStatus": "",
        "type": "ADJUDICATION",
        "verdict": "",
      },
      {
        "disposalQualifiers": "",
        "disposalQuantity": "    11122024          00",
        "disposalText": "",
        "disposalType": "2060",
        "type": "DISPOSAL",
      },
      {
        "committedOnBail": "N",
        "courtOffenceSequenceNumber": null,
        "locationOfOffence": "Not provided by Court",
        "offenceEndDate": "",
        "offenceEndTime": "",
        "offenceLocationFSCode": "02BJ",
        "offenceReason": "RT88191",
        "offenceReasonSequence": "",
        "offenceStartDate": "01012025",
        "offenceStartTime": "",
        "type": "ARREST",
      },
      {
        "hearingDate": "05122024",
        "numberOffencesTakenIntoAccount": "0000",
        "pleaStatus": "",
        "type": "ADJUDICATION",
        "verdict": "NON-CONVICTION",
      },
      {
        "disposalQualifiers": "",
        "disposalQuantity": "    11122024          00",
        "disposalText": "",
        "disposalType": "2061",
        "type": "DISPOSAL",
      },
    ],
    "courtCaseReferenceNumber": "97/1626/8395Q",
    "courtHouseName": "",
    "croNumber": null,
    "dateOfHearing": "05122024",
    "forceStationCode": "02YZ",
    "generatedPNCFilename": "",
    "hearingsAdjudicationsAndDisposals": [],
    "pendingCourtDate": null,
    "pendingCourtHouseName": null,
    "pendingPsaCourtCode": null,
    "pncCheckName": null,
    "pncIdentifier": null,
    "preTrialIssuesUniqueReferenceNumber": "01ZD/0303208",
    "psaCourtCode": null,
  },
}
`;
